bandit/bandit.go

```diff
@@ -178,21 +178,42 @@ func (o *BanditDialer) DialContext(ctx context.Context, network, addr string) (n
 // hasSucceedingDialer checks whether or not any of the given dialers is able to successfully dial our proxies
 func hasSucceedingDialer(dialers []Dialer) bool {
 	for _, dialer := range dialers {
-		if dialer.Succeeding() {
+		if dialer.ConsecFailures() == 0 && dialer.Successes() > 0 {
 			return true
 		}
 	}
 	return false
 }
 
-func (o *BanditDialer) chooseDialerForDomain(dialers []Dialer, network, addr string) (Dialer, int) {
+// hasNotFailing checks whether or not any of the given dialers are not explicitly failing
+func hasNotFailing(dialers []Dialer) bool {
+	for _, dialer := range dialers {
+		if dialer.ConsecFailures() == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (o *BanditDialer) chooseDialerForDomain(network, addr string) (Dialer, int) {
+	// Loop through the number of dialers we have and select the one that is best
+	// for the given domain.
 	chosenArm := o.bandit.SelectArm(rand.Float64())
-	dialer := dialers[chosenArm]
-	if dialer.SupportsAddr(network, addr) {
-		return dialer, chosenArm
+	notAllFailing := hasNotFailing(o.dialers)
+	for i := 0; i < (len(o.dialers) * 2); i++ {
+		dialer := o.dialers[chosenArm]
+		if (dialer.ConsecFailures() > 0 && notAllFailing) || !dialer.SupportsAddr(network, addr) {
+			// If the chosen dialer has consecutive failures and there are other
+			// dialers that are succeeding, we should choose a different dialer.
+			//
+			// If the chosen dialer does not support the address, we should also
+			// choose a different dialer.
+			chosenArm = differentArm(chosenArm, len(o.dialers))
+		} else {
+			break
+		}
```

Comment:
This suggestion is totally optional since the code already works as intended
```suggestion
                         continue
		}
	        break
```